tuf/client_rework/metadata_bundle.py

```diff
@@ -0,0 +1,464 @@
+                error type and content will contain more details.
+        """
+        self._bundle = {}  # type: Dict[str: Metadata]
+        self.reference_time = datetime.utcnow()
+        self._root_update_finished = False
+
+        # Load and validate the local root metadata
+        # Valid root metadata is required
+        logger.debug("Updating initial trusted root")
+        self.update_root(data)
+
+    def __getitem__(self, role: str) -> Metadata:
+        """Returns current Metadata for 'role'"""
+        return self._bundle[role]
+
+    def __len__(self) -> int:
+        """Returns number of Metadata objects in bundle"""
+        return len(self._bundle)
+
+    def __iter__(self) -> Iterator[Metadata]:
+        """Returns iterator over all Metadata objects in bundle"""
+        return iter(self._bundle)
+
+    # Helper properties for top level metadata
+    @property
+    def root(self) -> Optional[Metadata]:
+        """Current root Metadata or None"""
+        return self._bundle.get("root")
+
+    @property
+    def timestamp(self) -> Optional[Metadata]:
+        """Current timestamp Metadata or None"""
+        return self._bundle.get("timestamp")
+
+    @property
+    def snapshot(self) -> Optional[Metadata]:
+        """Current snapshot Metadata or None"""
+        return self._bundle.get("snapshot")
+
+    @property
+    def targets(self) -> Optional[Metadata]:
+        """Current targets Metadata or None"""
+        return self._bundle.get("targets")
+
+    # Methods for updating metadata
+    def update_root(self, data: bytes):
+        """Verifies and loads 'data' as new root metadata.
+
+        Note that an expired intermediate root is considered valid: expiry is
+        only checked for the final root in root_update_finished().
+
+        Args:
+            data: unverified new root metadata as bytes
+
+        Raises:
+            RepositoryError: Metadata failed to load or verify. The actual
+                error type and content will contain more details.
+        """
+        if self._root_update_finished:
+            raise RuntimeError(
+                "Cannot update root after root update is finished"
+            )
+        logger.debug("Updating root")
+
+        try:
+            new_root = Metadata.from_bytes(data)
+        except DeserializationError as e:
+            raise exceptions.RepositoryError("Failed to load root") from e
+
+        if new_root.signed.type != "root":
+            raise exceptions.RepositoryError(
+                f"Expected 'root', got '{new_root.signed.type}'"
+            )
+
+        if self.root is not None:
+            # We are not loading initial trusted root: verify the new one
+            if not verify_with_threshold(self.root, "root", new_root):
+                raise exceptions.UnsignedMetadataError(
+                    "New root is not signed by root", new_root.signed
+                )
+
+            if new_root.signed.version != self.root.signed.version + 1:
```

Comment:
Per the spec, this comes _after_ verifying with a threshold of keys from the new root metadata. The code is certainly nicer if it happens in this same `if`... do you think we should move it?

Reply:
You're correct... 

I think I'm happy with it being here (this feels like the spec doing things in the non-optimal order and the code happens to look nicer this way as well) but no strong opinions.